File: backend/arkham-api/library/Arkham/Xp.hs
{-# LANGUAGE NoFieldSelectors #-}

module Arkham.Xp where

import Arkham.Id
import Arkham.Prelude
import GHC.Records

data XpSource = XpFromVictoryDisplay | XpBonus | XpFromCardEffect
  deriving stock (Show, Ord, Eq, Generic, Data)
  deriving anyclass (ToJSON, FromJSON)

data XpDetail = XpDetail
  { source :: XpSource
  , sourceName :: Text
  , amount :: Int
  }
  deriving stock (Show, Ord, Eq, Generic, Data)
  deriving anyclass (ToJSON, FromJSON)

data XpEntry
  = AllGainXp {details :: XpDetail}
  | InvestigatorGainXp {investigator :: InvestigatorId, details :: XpDetail}
  | InvestigatorLoseXp {investigator :: InvestigatorId, details :: XpDetail}
  | {- | A non-XP campaign counter (Yig's Fury, ...) reported alongside the
    experience so the campaign log can attribute it. @tally@ is the i18n key
    for the counter's name, @details@ names the source and the amount, and
    @tallyOwner@ is 'Nothing' for a scenario-wide counter.
    -}
    TallyGained {tally :: Text, tallyOwner :: Maybe InvestigatorId, details :: XpDetail}
  | {- | The same counter going the other way. @details.amount@ is the
    magnitude, always positive.
    -}
    TallyLost {tally :: Text, tallyOwner :: Maybe InvestigatorId, details :: XpDetail}
  deriving stock (Show, Ord, Eq, Generic, Data)
  deriving anyclass (ToJSON, FromJSON)

instance HasField "amount" XpEntry Int where
  getField = \case
    AllGainXp details -> details.amount
    InvestigatorGainXp _ details -> details.amount
    InvestigatorLoseXp _ details -> details.amount
    TallyGained _ _ details -> details.amount
    TallyLost _ _ details -> details.amount

newtype XpBreakdown = XpBreakdown [XpEntry]
  deriving newtype (Monoid, Semigroup, Show, Ord, Eq, ToJSON, FromJSON)
  deriving stock Data

File: frontend/src/arkham/types/Xp.ts
import * as JsonDecoder from 'ts.data.json';
import { CampaignStep, campaignStepDecoder} from '@/arkham/types/CampaignStep';

export type XpSource = 'XpFromVictoryDisplay' | 'XpBonus' | 'XpFromCardEffect'

export type XpDetail = {
  source : XpSource;
  sourceName: string;
  amount: number;
}

export type XpEntry
  = { tag: 'AllGainXp', details: XpDetail }
  | { tag: 'InvestigatorGainXp', investigator: string, details: XpDetail }
  | { tag: 'InvestigatorLoseXp', investigator: string, details: XpDetail }

export type XpBreakdownStep = {
  step: CampaignStep
  investigators: string[]
  entries: XpEntry[]
}

export type LegaxyXpBreakdownStep = {
  step: CampaignStep
  entries: XpEntry[]
}

export type XpBreakdown = XpBreakdownStep[];

export const xpSourceDecoder = JsonDecoder.oneOf<XpSource>(
  [
    JsonDecoder.literal('XpFromVictoryDisplay'),
    JsonDecoder.literal('XpBonus'),
    JsonDecoder.literal('XpFromCardEffect'),
  ],
  'XpSource'
);

export const xpDetailDecoder = JsonDecoder.object<XpDetail>(
  {
    source: xpSourceDecoder,
    sourceName: JsonDecoder.string(),
    amount: JsonDecoder.number()
  },
  'XpDetail',
);

export const xpEntryDecoder = JsonDecoder.oneOf<XpEntry>(
  [
    JsonDecoder.object<XpEntry>(
      {
        tag: JsonDecoder.literal('AllGainXp'),
        details: xpDetailDecoder
      },
      'AllGainXp'
    ),
    JsonDecoder.object<XpEntry>(
      {
        tag: JsonDecoder.literal('InvestigatorGainXp'),
        investigator: JsonDecoder.string(),
        details: xpDetailDecoder
      },
      'InvestigatorGainXp'
    ),
    JsonDecoder.object<XpEntry>(
      {
        tag: JsonDecoder.literal('InvestigatorLoseXp'),
        investigator: JsonDecoder.string(),
        details: xpDetailDecoder
      },
      'InvestigatorLoseXp'
    ),
  ],
  'XpEntry'
);

export const xpBreakdownStepDecoder = JsonDecoder.object<XpBreakdownStep>(
  {
    step: campaignStepDecoder,
    investigators: JsonDecoder.array(JsonDecoder.string(), 'string[]'),
    entries: JsonDecoder.array(xpEntryDecoder, 'XpEntry[]'),
  },
  'XpBreakdownStep'
);

export const legacyXpBreakdownStepDecoder = JsonDecoder.object<LegaxyXpBreakdownStep>(
  {
    step: campaignStepDecoder,
    entries: JsonDecoder.array(xpEntryDecoder, 'XpEntry[]'),
  },
  'XpBreakdownStep'
);

export const xpBreakdownDecoder =
  JsonDecoder.oneOf<XpBreakdownStep[]>([
    JsonDecoder.array(xpBreakdownStepDecoder, 'XpBreakdown'),
    JsonDecoder.array(legacyXpBreakdownStepDecoder, 'LegacyXpBreakdown').map(legacy => legacy.map(step => ({ ...step, investigators: [] } as XpBreakdownStep)))
  ], 'XpBreakdown');

